src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java

```diff
@@ -216,13 +307,53 @@ private PaymentHandleResponse executeWithExceptionallyHandling(String paymentId,
+                            return saveResponseToInterfaceInteraction(paymentId, paymentIdType, restException)
+                                    .toCompletableFuture().join();
                         }
                     }
-                    return PaymentHandleResponse.of400BadRequest(
-                            format("%s [%s] can't be processed, see the logs", paymentIdType, paymentId));
-                }).toCompletableFuture().join();
+                    CompletionStage<PaymentHandleResponse> paymentHandleResponseStage = CompletableFuture.completedFuture(PaymentHandleResponse.of400BadRequest(
+                            format("%s [%s] can't be processed, details: [%s]", paymentIdType, paymentId, throwable.getMessage())));
+                    return paymentHandleResponseStage.toCompletableFuture().join();
+                });
+        return exceptionally.toCompletableFuture().join();
+    }
+
+    private CompletionStage<PaymentHandleResponse> saveResponseToInterfaceInteraction(@Nullable String paymentId,
+                                                                                      @Nullable String paymentIdType,
+                                                                                      @Nonnull PayPalRESTException restException) {
+        CompletionStage<String> ctpPaymentIdStage = CompletableFuture.completedFuture(paymentId);
+        if (PAYPAL_PLUS_PAYMENT_ID.equals(paymentIdType)) {
+            // if it's paypal plus payment id and not ctp payment id,
+            // then fetch ctp payment to get ctp payment id
+            ctpPaymentIdStage = this.ctpFacade.getPaymentService()
+                    .getByPaymentMethodAndInterfaceId(PAYPAL_PLUS, paymentId)
+                    .thenApply(ctpPayment -> ctpPayment.map(Resource::getId).orElse(null));
+        }
 
+        CompletionStage<PaymentHandleResponse> paymentHandleResponseStage = ctpPaymentIdStage
+                .thenCompose(ctpPaymentId -> {
+                    AddInterfaceInteraction action = createAddInterfaceInteractionAction(restException.getDetails(), RESPONSE);
+                    return ctpFacade.getPaymentService().updatePayment(ctpPaymentId, Collections.singletonList(action))
+                            .thenApply(ignore -> {
+                                return PaymentHandleResponse.ofHttpStatusAndErrorMessage(HttpStatus.valueOf(restException.getResponsecode()),
+                                        format("%s=[%s] can't be processed, details: [%s]", paymentIdType, paymentId, restException.getMessage()));
+                            });
+                });
+        // todo: don't join here, but rather return completion stage
+        return paymentHandleResponseStage;
+    }
+
+    private Optional<String> getCtpPaymentId(@Nonnull Cart cartWithPaymentsExpansion,
+                                   @Nonnull String paypalPlusPaymentId) {
+        List<Reference<io.sphere.sdk.payments.Payment>> payments
```

Comment:
i'd recommend more straightforward way, like this:

```java
return Optional.of(cartWithPaymentsExpansion)
        .map(Cart::getPaymentInfo)
        .map(PaymentInfo::getPayments)
        .map(List::stream)
        .flatMap(paymentsStream -> paymentsStream
                .filter(paymentReference -> paymentReference.getObj() != null) // you could join these 2 lines to one `map` line with exception, if you really need _MissingExpansionException_ exception here. But i don't think so.
                .map(Reference::getObj)
                .filter(payment -> paypalPlusPaymentId.equals(payment.getInterfaceId()))
                .findFirst())
        .map(io.sphere.sdk.payments.Payment::getId);
```

Reply:
I'm not sure if it's more straightforward, I think the nested `flatMap` makes it harder to read. And also if I want to throw `MissingExpansionException`, it's hard to do.

Reply:
the `flatMap` is exactly to avoid nesting of optional objects. You could extract the content of `flatMap` lambda to a dedicated method - then it will be even more readable. 

And if you really need this exception here (do you?) - you could use the comment i put in the code snippet:

>you could join these 2 lines to one `map` line with exception, if you really need _MissingExpansionException_ exception here. But i don't think so.

Still i'm confused why you check the `getObj()` here and why only the first one? The whole approach of this verification looks weird.